### PR TITLE
Feature/add meson support

### PR DESCRIPTION
**Pull request by lumurillo**
Add meson support to GstInference

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+if get_option('enable-profiling').enabled()
+  profiler_dep = dependency('libprofiler')
+  if(profiler_dep.found())
+    message('Profiling enabled: Building examples with profiling support.')
+    link_flags += ['-lprofiler']
+    c_flags += ['-DPROFILING']
+    # Update test test_deps to include profiler dependency
+    test_deps += [profiler_dep]
+  else
+    error('MESON_FAIL gperftools profiling library not found.')
```
Comment (mleivaf) [inline meson.build:50]: Is required to fail if gperftools does not exist?

File: ext/opencv/meson.build [hunk 2]
```diff
@@ -0,0 +1,98 @@
+gstinference_sources = [
+  'gstplugin.cc',
+  'gstclassificationoverlay.cc',
+  'gstinferenceoverlay.cc',
+  'gstdetectionoverlay.cc',
+  'gstembeddingoverlay.cc'
+]
+
+libopencv_headers = [
+  'opencv2/bgsegm.hpp',
```
Comment (michaelgruner) [inline ext/opencv/meson.build:10]: Do we really need al these headers?

File: ext/opencv/meson.build [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+libopencv4_headers = [
+  'opencv4/opencv2/calib3d.hpp',
+  'opencv4/opencv2/core.hpp',
+  'opencv4/opencv2/imgproc.hpp',
+  'opencv4/opencv2/objdetect.hpp',
+  'opencv4/opencv2/opencv.hpp',
+  'opencv4/opencv2/video.hpp',
+]
+
+gstopencv_cargs = ['-DGST_HAAR_CASCADES_DIR="@0@"']
```
Comment (michaelgruner) [inline ext/opencv/meson.build:28]: This is not necessary at all

File: ext/opencv/meson.build [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+opencv_dep = dependency('opencv', version : ['>= 3.0.0', '< 3.5.0'], required : false)
+opencv_found = opencv_dep.found()
+
+if opencv_found
+  foreach h : libopencv_headers
+    if not cxx.has_header(h)
+      message('Needed header "' + h + '" not found')
+      opencv_found = false
```
Comment (michaelgruner) [inline ext/opencv/meson.build:37]: let's break in case of an error

File: ext/opencv/meson.build [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  'opencv4/opencv2/video.hpp',
+]
+
+gstopencv_cargs = ['-DGST_HAAR_CASCADES_DIR="@0@"']
+
+opencv_dep = dependency('opencv', version : ['>= 3.0.0', '< 3.5.0'], required : false)
+opencv_found = opencv_dep.found()
+
+if opencv_found
+  foreach h : libopencv_headers
```
Comment (michaelgruner) [inline ext/opencv/meson.build:34]: Is the headers check really necessary?

File: ext/opencv/meson.build [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    endif
+  endif
+endif
+
+gst_opencv_all_sources = []
+gst_opencv_all_dependencies = [gst_base_dep, gst_video_dep, opencv_dep, gstopencv_dep]
+if gstinference_dep.found()
+  gst_opencv_all_sources += gstinference_sources
+  gst_opencv_all_dependencies += gstinference_dep
+endif
```
Comment (michaelgruner) [inline ext/opencv/meson.build:83]: What is all this? This has nothing to do with us. Please emulate the checks we are doing on our configure.ac in this meson file.

File: gst-libs/gst/opencv/meson.build [hunk 7]
```diff
@@ -0,0 +1,43 @@
+opencv_sources = [
+  'gstinferencebaseoverlay.cc'
+]
+
+opencv_headers = [
+  'gstinferencebaseoverlay.h'
+]
+
+gstinference_headers = [
+  'gstinferenceoverlay.h'
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:10]: What is this header? this doesn't exist. Does this build?

File: gst-libs/gst/opencv/meson.build [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+]
+
+opencv_dep = dependency('opencv', version : '>= 3.0.0', required : false)
+r2inference_dep = dependency('r2inference-0.0', required : true)
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:14]: Do we need to search for this dependency again?

File: gst-libs/gst/opencv/meson.build [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+opencv_headers = [
+  'gstinferencebaseoverlay.h'
+]
+
+gstinference_headers = [
+  'gstinferenceoverlay.h'
+]
+
+opencv_dep = dependency('opencv', version : '>= 3.0.0', required : false)
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:13]: We have very critical requirements with opencv versions. Please follow the same logic as in our configure.ac. Also, we need some macros declared for our code, specifically `OCV_VERSION_*`

File: gst-libs/gst/opencv/meson.build [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+r2inference_dep = dependency('r2inference-0.0', required : true)
+
+if not opencv_dep.found()
+  opencv_dep = dependency('opencv4', version : '>= 4.0.0', required : false)
+endif
+if opencv_dep.found()
+	opencv_all_sources = opencv_sources
+	opencv_all_headers = opencv_headers
+	opencv_all_dependencies = [gst_base_dep, gst_video_dep, opencv_dep]
+	if r2inference_dep.found()
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:23]: We can't continue if there's no r2inference

File: gst-libs/gst/opencv/meson.build [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+	  opencv_all_dependencies += r2inference_dep
+	endif
+  gstopencv = library('gstinferencebaseoverlay-@0@'.format(api_version),
+    opencv_all_sources,
+    c_args : c_args + ['-DBUILDING_GST_OPENCV']+ ['-DBUILDING_GST_INFERENCE'],
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:28]: We are not building gst-opencv

File: gst-libs/gst/opencv/meson.build [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    cpp_args: cpp_args + ['-DBUILDING_GST_OPENCV']+ ['-DBUILDING_GST_INFERENCE'],
+    include_directories : [configinc, inference_inc_dir],
+    version : '0.0.0', # TODO verify this version number
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:31]: don't leave dirty code behind

File: ext/r2inference/meson.build [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+	'gstinceptionv3.c',
+	'gstinceptionv4.c',
+	'gstinference.c',
+	'gstmobilenetv2.c',
+	'gstresnet50v1.c',
+	'gsttinyyolov2.c',
+	'gsttinyyolov3.c'
+]
+
+gstinference_headers = [
```
Comment (michaelgruner) [inline ext/r2inference/meson.build:14]: What are these used for?

File: meson.build [hunk 14]
```diff
@@ -0,0 +1,223 @@
+project('GStreamer Inference', ['c', 'cpp'],
+  version : '0.9.1.1',
```
Comment (michaelgruner) [inline meson.build:2]: This does not match our current version

File: meson.build (hunk 1 above)
Comment (lumurillo) [reply meson.build:50]: No, I changed it to message

File: ext/opencv/meson.build (hunk 2 above)
Comment (lumurillo) [reply ext/opencv/meson.build:10]: No, just removed.

File: ext/opencv/meson.build (hunk 3 above)
Comment (lumurillo) [reply ext/opencv/meson.build:28]: Removed.

File: ext/opencv/meson.build (hunk 4 above)
Comment (lumurillo) [reply ext/opencv/meson.build:37]: This **for** statement was removed since the headers were not needed.

File: ext/opencv/meson.build (hunk 5 above)
Comment (lumurillo) [reply ext/opencv/meson.build:34]: No, removed.

File: ext/opencv/meson.build (hunk 6 above)
Comment (lumurillo) [reply ext/opencv/meson.build:83]: Removed, no needed.

File: gst-libs/gst/opencv/meson.build (hunk 7 above)
Comment (lumurillo) [reply gst-libs/gst/opencv/meson.build:10]: Removed.

File: gst-libs/gst/opencv/meson.build (hunk 8 above)
Comment (lumurillo) [reply gst-libs/gst/opencv/meson.build:14]: Moved to the main meson.build.

File: gst-libs/gst/opencv/meson.build (hunk 10 above)
Comment (lumurillo) [reply gst-libs/gst/opencv/meson.build:23]: Added error here.

File: gst-libs/gst/opencv/meson.build (hunk 9 above)
Comment (lumurillo) [reply gst-libs/gst/opencv/meson.build:13]: Added those macros in the main mason.build

File: gst-libs/gst/opencv/meson.build (hunk 11 above)
Comment (lumurillo) [reply gst-libs/gst/opencv/meson.build:28]: Removed.

File: gst-libs/gst/opencv/meson.build (hunk 12 above)
Comment (lumurillo) [reply gst-libs/gst/opencv/meson.build:31]: This is the correct version, removed the comment

File: ext/r2inference/meson.build (hunk 13 above)
Comment (lumurillo) [reply ext/r2inference/meson.build:14]: Removed

File: meson.build (hunk 14 above)
Comment (lumurillo) [reply meson.build:2]: Changed to 0.10.0.1

File: gst-libs/gst/opencv/meson.build [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+if opencv_dep.found()
+	opencv_all_sources = opencv_sources
+	opencv_all_headers = opencv_headers
+	opencv_all_dependencies = [gst_base_dep, gst_video_dep, opencv_dep]
+	if r2inference_dep.found()
+	  opencv_all_dependencies += r2inference_dep
+  else
+    error('R2Inference dependency not found.')
+	endif
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:17]: Indentation

File: gst-libs/gst/opencv/meson.build [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+  gstopencv = library('gstinferencebaseoverlay-@0@'.format(api_version),
+    opencv_all_sources,
+    c_args : c_args + ['-DBUILDING_GST_INFERENCE'],
+    cpp_args: cpp_args + ['-DBUILDING_GST_INFERENCE'],
+    include_directories : [configinc, inference_inc_dir],
+    version : '0.0.0',
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:24]: Match our version (don't hardcode)

File: gst-libs/meson.build [hunk 17]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (michaelgruner) [inline gst-libs/meson.build:1]: Extra line

File: ext/opencv/meson.build [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+  'gstplugin.cc',
+  'gstclassificationoverlay.cc',
+  'gstinferenceoverlay.cc',
+  'gstdetectionoverlay.cc',
+  'gstembeddingoverlay.cc'
+]
+
+opencv_all_sources = []
+opencv_all_dependencies = [gst_base_dep, gst_video_dep, opencv_dep, gstopencv_dep]
+if gstinference_dep.found()
```
Comment (michaelgruner) [inline ext/opencv/meson.build:11]: I don't feel this check is necessary since its an internal dependency.

File: ext/opencv/meson.build [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+  gstopencv = library('gstinferenceoverlayplugin',
+    opencv_all_sources,
+    cpp_args : cpp_args,
+    include_directories : [configinc, inference_inc_dir],
+    dependencies : opencv_all_dependencies,
+    install : true,
+    install_dir : plugins_install_dir,
+  )
+else
+  error('OpenCV dependencies were not found.')
```
Comment (michaelgruner) [inline ext/opencv/meson.build:28]: In autotools we don't fail if the system doesn't have opencv installed. We print a warning and continue without building these elements.

File: gst-libs/gst/opencv/meson.build [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+    dependencies : opencv_all_dependencies,
+  )
+
+  gstopencv_dep = declare_dependency(link_with: gstopencv,
+    include_directories : [inference_inc_dir],
+    dependencies : opencv_all_dependencies)
+
+  install_headers(opencv_all_headers, subdir : 'gstreamer-1.0/gst/opencv')
+else
+  error('OpenCV dependencies were not found.')
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:35]: We don't fail if we don't find OpenCV. We just print a warning and continue without building this library.

File: ext/opencv/meson.build [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+endif
```
Comment (michaelgruner) [inline ext/opencv/meson.build:29]: We need to update the `plugins` array

File: gst-libs/gst/opencv/meson.build (hunk 15 above)
Comment (lumurillo) [reply gst-libs/gst/opencv/meson.build:17]: Fixed

File: gst-libs/meson.build (hunk 17 above)
Comment (lumurillo) [reply gst-libs/meson.build:1]: Removed

File: ext/opencv/meson.build (hunk 18 above)
Comment (lumurillo) [reply ext/opencv/meson.build:11]: Removed

File: ext/opencv/meson.build (hunk 19 above)
Comment (lumurillo) [reply ext/opencv/meson.build:28]: Changed it to warning message

File: gst-libs/gst/opencv/meson.build (hunk 20 above)
Comment (lumurillo) [reply gst-libs/gst/opencv/meson.build:35]: Changed it to warning message

File: ext/opencv/meson.build (hunk 21 above)
Comment (lumurillo) [reply ext/opencv/meson.build:29]: Added to pllugins array

Comment (michaelgruner) [reply ext/opencv/meson.build:25]: I don't feel this means anything to the end user. Maybe something like: `OpenCV not found in this system. Overlay plug-ins will not be build.`

File: gst-libs/gst/opencv/meson.build [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    dependencies : opencv_all_dependencies,
+  )
+
+  gstopencv_dep = declare_dependency(link_with: gstopencv,
+    include_directories : [inference_inc_dir],
+    dependencies : opencv_all_dependencies)
+
+  install_headers(opencv_all_headers, subdir : 'gstreamer-1.0/gst/opencv')
+else
+  warning('OpenCV dependencies were not found for gstinferencebaseoverlay.')
```
Comment (michaelgruner) [inline gst-libs/gst/opencv/meson.build:35]: Again, I recommend a more friendly message. Something like `OpenCV was not found in this system. Base overlay utils will not be built.`

File: gst-libs/gst/r2inference/meson.build [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+  )
+
+  gstinference_dep = declare_dependency(link_with: gstinference,
+    include_directories : [inference_inc_dir],
+    dependencies : [gst_video_dep, r2inference_dep])
+
+  install_headers(gstinference_headers, subdir : 'gstreamer-1.0/gst/r2inference')
+  pkgconfig.generate(gstinference, install_dir : plugins_pkgconfig_install_dir)
+else
+  error('r2inference dependencies were not found.')
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/meson.build:51]: Friendlier message: `r2inference not found. Please install before proceeding with the installation`.

File: gst-libs/gst/opencv/meson.build (hunk 10 above)
Comment (michaelgruner) [reply gst-libs/gst/opencv/meson.build:13]: Please double check if r2inference is needed. I don't believe so, but I may be wrong.

File: gst-libs/gst/meson.build [hunk 24]
```diff
@@ -0,0 +1,2 @@
+subdir('opencv')
+subdir('r2inference')
```
Comment (michaelgruner) [inline gst-libs/gst/meson.build:2]: Build first r2i

File: gst-libs/gst/r2inference/meson.build [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+	'gsttflite.h',
+	'gstvideoinference.h'
+]
+
+if r2inference_dep.found()
+  gstinference = library('gstinference-1.0',
+    gstinference_sources,
+    c_args : c_args,
+    include_directories : [configinc, inference_inc_dir],
+    version : version_arr[0],
```
Comment (michaelgruner) [inline gst-libs/gst/r2inference/meson.build:39]: If we add version here, we need to add it to BaseOverlay as well.

Comment (lumurillo) [reply gst-libs/gst/r2inference/meson.build:39]: The base overlay already has this version

**Pull request merged**